Volume offset control service and client

Thalley opened the pull request:
This PR implements the Volume Offset Control Service (VOCS) server and client. 

The shell has purposely not been modified to control this server, as the service is a secondary service and only has purpose in the context of a primary service. 

This PR is the first in a series of 3 to add the Volume Control Service server and client:
1) Add Volume Offset Control Service (VOCS) server and client (This)
2) Add Audio Input Control Service (AICS) server and client (https://github.com/zephyrproject-rtos/zephyr/pull/32785)
3) Add Volume Control Service (VCS) server and client (TBD)

The reason why VOCS and AICS should be implemented first, is that VCS (as a primary service) uses these two secondary services, and having them implemented first will make it easier to merge.

Thalley commented:
At the time of this comment, the first 2 commits, https://github.com/zephyrproject-rtos/zephyr/pull/31893/commits/0a102311afcc08fd7fa23cc4b26199f100331588 and https://github.com/zephyrproject-rtos/zephyr/pull/31893/commits/b628975de1bf6d7fd858234310bb6eeb6505d840, are commits from https://github.com/zephyrproject-rtos/zephyr/pull/31182 that are necessary to build. Once https://github.com/zephyrproject-rtos/zephyr/pull/31182 have been merged, these commits will be removed/ignored from this PR.

asbjornsabo reviewed (changes requested):
Some comments for consideration.
The question of the size of the audio location parameters has been submitted as an errata to the spec, and must be handled by GAWG

Thalley commented on include/bluetooth/audio/vocs.h:
```diff
@@ -0,0 +1,291 @@
+
+
+/** @brief Callback function for writes.
+ *
```
Named to set_offset_cb, as that is the only use now (had more uses previously).

Thalley commented on include/bluetooth/audio/vocs.h:
```diff
@@ -0,0 +1,291 @@
+ *                       or NULL to read local server value.
+ *  @param inst          Pointer to the Volume Offset Control Service instance.
+ *
+ *  @return 0 on success, GATT error value on fail.
```
Added clear description of that.

Thalley commented on include/bluetooth/audio/vocs.h:
```diff
@@ -0,0 +1,291 @@
+int bt_vocs_init(struct bt_vocs *vocs, struct bt_vocs_init *init);
+
+/** @brief Get a free instance of Volume Offset Control Service from the pool.
+ *
```
Added "service"

Thalley commented on subsys/bluetooth/audio/vocs.c:
```diff
@@ -0,0 +1,464 @@
+	BT_GATT_CCC(offset_state_cfg_changed, \
+		    BT_GATT_PERM_READ | BT_GATT_PERM_WRITE_ENCRYPT), \
+	BT_GATT_CHARACTERISTIC(BT_UUID_VOCS_LOCATION, \
+			       BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY, \
```
That is updated in `bt_vocs_init`

Thalley commented on subsys/bluetooth/audio/vocs.c:
```diff
@@ -0,0 +1,464 @@
+			       BT_GATT_PERM_WRITE_ENCRYPT, \
+			       NULL, write_vocs_control, &_vocs), \
+	BT_GATT_CHARACTERISTIC(BT_UUID_VOCS_DESCRIPTION, \
+			       BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY, \
```
That is updated in `bt_vocs_init`

Thalley commented on include/bluetooth/audio/vocs.h:
```diff
@@ -0,0 +1,291 @@
+ *
+ *  @return 0 if success, ERRNO on failure.
+ */
+int bt_vocs_init(struct bt_vocs *vocs, struct bt_vocs_init *init);
```
Added `param` to the struct

Thalley commented on subsys/bluetooth/audio/vocs.c:
```diff
@@ -0,0 +1,464 @@
+	vocs->srv.state.offset = init->offset;
+	for (int i = 0; i < ARRAY_SIZE(vocs_insts); i++) {
+		vocs_insts[i].service_p = &vocs_service_list[i];
+	}
```
Agreed. Move the code for all instances (just this loop) to a function and added a static check for it.

Thalley commented on subsys/bluetooth/audio/vocs.c:
```diff
@@ -0,0 +1,464 @@
+	/* Iterate over the attributes in AICS (starting from i = 1 to skip the
+	 * service declaration) to find the BT_UUID_VOCS_DESCRIPTION or
+	 * BT_UUID_VOCS_LOCATION and update the characteristic value (at [i]),
+	 * update those with the write permission and callback, and also update
```
Removed "those"; I still think "update" should be there.

Thalley commented on subsys/bluetooth/audio/vocs.c:
```diff
@@ -0,0 +1,464 @@
+#if defined(CONFIG_BT_VOCS) || defined(CONFIG_BT_VOCS_CLIENT)
+
+int bt_vocs_state_get(struct bt_conn *conn, struct bt_vocs *inst)
+{
```
I've added some `CHECKIF`s for that

Thalley commented on subsys/bluetooth/audio/vocs_client.c:
```diff
@@ -0,0 +1,642 @@
+			       inst->cli.state.offset,
+			       inst->cli.state.change_counter);
+
+			/* clear busy flag to reuse function */
```
Added a comment where we handle the invalid change counter. 

Where would additional information about the abstraction of the state counter be described? Top of the vocs.h file?

asbjornsabo replied:
Sounds fine to me - something along the lines of "Note that this API abstracts away the state counter described in the Volume Offset Control Service specifiction.  If required, the client will automatically refresh its server state information before executing the write request. ?

asbjornsabo commented on include/bluetooth/audio/vocs.h:
```diff
@@ -0,0 +1,291 @@
+ */
+struct bt_vocs_init {
+	/** Audio Location bitmask */
+	uint8_t location;
```
FYI: The spec author has agreed, a fix to the service spec is on its way.

Thalley committed:
Bluetooth: Audio: Volume Offset Control Service and Client

This commit implements the secondary service
Volume Offset Control Service (VOCS) server and client.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

asbjornsabo reviewed (approved):
All my comments have been resolved to my satisfaction.
(I have not checked for other changes.)

Thalley commented:
> All my comments have been resolved to my satisfaction.
> (I have not checked for other changes.)

Thanks @asbjornsabo - Now we just need 1 more review :)

jhedberg commented on include/bluetooth/audio/vocs.h:
```diff
@@ -0,0 +1,304 @@
+ *                     For notifications, this will always be 0.
+ * @param description  The description as an UTF-8 encoded string.
+ */
+typedef void (*bt_vocs_description_cb_t)(
+	struct bt_conn *conn, struct bt_vocs *inst, int err, char *description);
```
I guess it's subjective, but I think it'd look nicer if you had included some function parameters on the first line, done the split after a parameter (i.e. after a `,` character). Same for the other function pointer typedefs

Thalley replied:
Agreed. This was previously impossible due to the 80 line length (as some single parameters couldn't not fit on the line if starting at `(`, but it's possible now.

jhedberg commented on subsys/bluetooth/audio/vocs_client.c:
```diff
@@ -0,0 +1,668 @@
+		    vocs_insts[i].cli.end_handle >= handle) {
+			return &vocs_insts[i];
+		}
+	}
```
Empty line after `}` (unless the next line is also `}`. I think I saw at least one other place in the PR with the same issue. (minor thing, I know, but something we've tried to be consistent with within the Bluetooth subsystem)

Thalley replied:
Agreed. This, imo, is however something a linter should handle for us, and not something you, as a reviewer, should need to deal with. Whatever linting we have that require empty line after a variable declaration: Can't that do this as well?

jhedberg replied:
Probably, but I'm not sure this rule is followed much outside of the Bluetooth subsystem, so that might make it hard to enforce it on others. Linux checkpatch has support for subsystem-specific quirks, but I don't think we want to go that route with Zephyr. There are all sorts of small personal preferences which I feel like they help readability, but probably not worth codifying anywhere. Another which I've stopped pointing out about is the "reverse x-mas tree" rule for variable declarations where the shortest lines ones come last 😄 

Even for this curly-brace rule, I'm not sure it applies in all cases, since the most important thing about empty lines is to group things which logically belong together into a single block. There could be cases where the line immediately following a block is so strongly related that adding the empty line detracts from easy understanding of the code. And that again means that automating this isn't really possible.

Thalley replied:
I agree :) My personal opinion is that any code style things (such as empty lines after variable declaration, or code blocks) should be handled by a linter. A lot of PR reviews become longer due to such comments (although it's generally good that they are there!), which could have been fixed by the uploader before assigning reviewers :) 

But as you also say, it's hard to apply such rules across the entirety of Zephyr as there are so many code styles applied in different subsystems.

jhedberg reviewed (changes requested):
Mostly coding style related comments, but also one functional issue regarding the meaning of  `data == NULL` in callbacks.

Thalley commented on subsys/bluetooth/audio/vocs_client.c:
```diff
@@ -0,0 +1,668 @@
+		return BT_GATT_ITER_STOP;
+	}
+
+	if (data) {
```
From gatt.h for `bt_gatt_notify_func_t`: 
```
@param data Attribute value data. If NULL then subscription was removed.
```
I can change the checks from `data == NULL` to `len == 0` as they (hopefully) always are paired like that. 

w.r.t. doing 

```
if (!cond) { 
    return;
}
```

I think that's a fine idea as it reduces indentation. 

Though, to be fair, I've received opposed review comments on some of my other PR, where people dislike have returns in the middle of functions, so that's subjective :D But I'll update them to be `if(!cond)` as that is my personal preference as well.

jhedberg replied:
I didn't realise NULL had special "subscription removed" meaning, rather than an actual indication/notification with empty payload. In that case checking for NULL is fine, but I'd do it the "early return" way along with a code comment about subscription removal.

jhedberg replied:
Regarding preferences for single function return point, I think those might originate in some MISRA-C -related context, since there exists coding style requirements which state that every function must have a single point of return (I don't agree with it and I don't think it will ever be applied tree-wide).

asbjornsabo replied:
Since the data pointer being NULL is the one specified in the documentation, I would rather trust that than trusting the length is also zero (the latter has not been specified).

Thalley commented on subsys/bluetooth/audio/vocs_client.c:
```diff
@@ -0,0 +1,668 @@
+	BT_DBG("Inst %p: err: 0x%02X", inst, err);
+	inst->cli.busy = false;
+
+	if (data) {
```
See https://github.com/zephyrproject-rtos/zephyr/pull/31893#discussion_r597681691

Thalley replied:
Actually, since this is for a read: If the client reads a characteristic, and the value returned by the GATT server has length 0, the data pointer will be NULL (and length as well). It's perfectly valid for the server to return 0 bytes from a read.

jhedberg commented on subsys/bluetooth/audio/vocs_client.c:
```diff
@@ -0,0 +1,689 @@
+		return BT_GATT_ITER_STOP;
+	}
+
+	if (!data || !length) {
```
Are you sure `length == 0` always means unsubscription, and that it isn't e.g. possible for the remote to send an empty payload which would result with `data != NULL && length == 0`?

Thalley replied:
You are completely correct here :) Removed the BT_DBG, but I guess we can't do much else here.

Thalley commented on include/bluetooth/audio/vocs.h:
```diff
@@ -0,0 +1,304 @@
+ * @ingroup bluetooth
+ * @{
+ *
+ * The Volume Offset Control Service is a secondary service, and as such should not be used own its
```
I actually don't have any auto-split in my editor, so any such changes I've done has been intentional. Personally I don't find liken of length 100 to be any less readable, so it is, as you say, subjective :) 

From: https://www.phoronix.com/scan.php?page=news_item&px=Linux-Kernel-Deprecates-80-Col

> Yes, staying withing 80 columns is certainly still _preferred_. But it's not the hard limit that the checkpatch warnings imply, and other concerns can most certainly dominate

So from the Linux Kernel point of view, 80 should still be preferred. Going forwards I'll try to keep it at max 80, unless to avoid ugly breaks :)

jhedberg commented on subsys/bluetooth/audio/vocs_client.c:
```diff
@@ -0,0 +1,688 @@
+		inst->cli.read_params.handle_count = 1;
+		inst->cli.read_params.single.handle = inst->cli.state_handle;
+
+		cb_err = bt_gatt_read(conn, &inst->cli.read_params);
+
+		if (cb_err) {
```
So here the empty line really shouldn't be there. Assigning to a variable and checking for the variable's value is considered to be a single logical chunk of code.

jhedberg commented on subsys/bluetooth/audio/vocs_client.c:
```diff
@@ -0,0 +1,688 @@
+		memcpy(desc, data, length);
+		desc[length] = '\0';
+		BT_DBG("Output description: %s", log_strdup(desc));
+	} else {
+		BT_DBG("Invalid location");
+		cb_err = BT_ATT_ERR_UNLIKELY;
+	}
```
What does `data == NULL` mean here? What is the stack trying to tell us?

Thalley replied:
That there is no data; i.e. the GATT server sent invalid data on our read. This is application specific (some characteristics may be of length 0, so we cannot do anything about this below the application level).

jhedberg commented on include/bluetooth/audio/vocs.h:
```diff
@@ -0,0 +1,302 @@
+ *
+ * @param conn         Connection to peer device, or NULL if local server read.
+ * @param inst         The instance pointer.
+ * @param err          Error value. 0 on success, GATT error or ERRNO on fail.
+ *                     For notifications, this will always be 0.
```
How do you differentiate between GATT errors and errno? Positive value is GATT, negative is -errno? If so, please document it clearly. Same for all other callbacks with a similar parameter.

Thalley replied:
Your assumption is correct: 0 = success, > 0 GATT and < 0 ERRNO. Updated.

jhedberg commented on subsys/bluetooth/audio/vocs_client.c:
```diff
@@ -0,0 +1,688 @@
+			 */
+			sub_params->ccc_handle = attr->handle + 2;
+			sub_params->notify = vocs_client_notify_handler;
+			err = bt_gatt_subscribe(conn, sub_params);
+
+			if (err) {
```
Remove the empty line (assignment to variable and testing for the value)

jhedberg commented on subsys/bluetooth/audio/vocs_client.c:
```diff
@@ -0,0 +1,688 @@
+		return -EPERM;
+	}
+
+	err = bt_gatt_write_without_response(conn, inst->cli.location_handle,
+					     &location, sizeof(location), false);
+
+	return err;
```
Seems like you can probably remove `err` completely and just do `return bt_gatt_write_without_response(...);`

jhedberg commented on subsys/bluetooth/audio/vocs_client.c:
```diff
@@ -0,0 +1,688 @@
+	inst->cli.read_params.single.offset = 0U;
+
+	err = bt_gatt_read(conn, &inst->cli.read_params);
+	if (!err) {
+		inst->cli.busy = true;
+	}
+	return err;
```
Here I wouldn't necessarily complain about the missing empty line after `}` since it's kind of a singular chunk of code, but you're actually not being consistent with yourself: there are other places such as the end of `bt_vocs_client_location_get()`, `bt_vocs_client_state_set()` and `bt_vocs_client_description_get()` where you *do* have the empty line, so I'd suggest to go with the same style here.

Thalley replied:
Agreed; tried to update it after your last comment regarding this. Must've missed this one :)

jhedberg commented on subsys/bluetooth/audio/vocs_internal.h:
```diff
@@ -0,0 +1,86 @@
+#ifndef ZEPHYR_INCLUDE_BLUETOOTH_AUDIO_VOCS_INTERNAL_
+#define ZEPHYR_INCLUDE_BLUETOOTH_AUDIO_VOCS_INTERNAL_
+#include <zephyr/types.h>
+
+
```
Remove one of the empty lines

jhedberg reviewed (changes requested):
A few more coding style comments, hopefully you have patience with me being pedantic about this stuff 😄

Thalley commented:
> A few more coding style comments, hopefully you have patience with me being pedantic about this stuff

Nah, it's great to get some fresh eyes on this :D

jhedberg reviewed (approved):
Looks good enough to me now thanks.

One thing which I don't feel strong enough to "request changes" anymore, that you might still want to do:

I wouldn't capitalize ERRNO since it's not an acronym or a literal API reference. It's referring to POSIX error numbers, and some POSIX APIs use the global (and thread unsafe) lower-case `errno` variable to store such values.

Thalley commented:
> Looks good enough to me now thanks.
> 
> One thing which I don't feel strong enough to "request changes" anymore, that you might still want to do:
> 
> I wouldn't capitalize ERRNO since it's not an acronym or a literal API reference. It's referring to POSIX error numbers, and some POSIX APIs use the global (and thread unsafe) lower-case `errno` variable to store such values.

That's a good point. I changed that too.

joerchan commented on subsys/bluetooth/audio/vocs_internal.h:
```diff
@@ -0,0 +1,85 @@
+#include <zephyr/types.h>
+
+#if defined(CONFIG_BT_VOCS)
+#define VOCS_MAX_DESC_SIZE CONFIG_BT_VOCS_MAX_OUTPUT_DESCRIPTION_SIZE
```
This header file is using both `bt_vocs` and `vocs` as the prefix, they should probably be all `bt_vocs`.

joerchan commented on subsys/bluetooth/audio/vocs.c:
```diff
@@ -0,0 +1,482 @@
+	if (len != sizeof(inst->location)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
```
the input argument offset is not checked or used. Should we return invalid offset if it is non-zero?

joerchan commented on subsys/bluetooth/audio/vocs.c:
```diff
@@ -0,0 +1,482 @@
+		inst->state.change_counter++;
+		BT_DBG("New state: offset %d, counter %u",
+		       inst->state.offset, inst->state.change_counter);
+		bt_gatt_notify_uuid(NULL, BT_UUID_VOCS_STATE, inst->service_p->attrs,
```
Should there be any handling of the return value here?

Thalley replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/31893#discussion_r598755623

joerchan commented on subsys/bluetooth/audio/vocs.c:
```diff
@@ -0,0 +1,482 @@
+				 const void *buf, uint16_t len, uint16_t offset, uint8_t flags)
+{
+	struct vocs_server *inst = attr->user_data;
+
```
return error if offset is non-zero?

joerchan commented on subsys/bluetooth/audio/vocs.c:
```diff
@@ -0,0 +1,482 @@
+		memcpy(inst->output_desc, buf, len);
+		inst->output_desc[len] = '\0';
+
+		bt_gatt_notify_uuid(NULL, BT_UUID_VOCS_DESCRIPTION, inst->service_p->attrs,
```
Check return value?

Thalley replied:
What do you suggest we do in case of error? Logging with BT_DBG or BT_WARN? 

There's not really anything besides logging we can do as the server.

joerchan replied:
I was mostly asking :) I suggest adding an explicit ignore of the return value.
`(void)bt_gatt_notify_uuid(...)`

joerchan commented on subsys/bluetooth/audio/vocs.c:
```diff
@@ -0,0 +1,482 @@
+		return -EINVAL;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_VOCS_CLIENT) && conn) {
```
Can the application be both client and server? And would that cause a problem when you call location get and you get the client always?

Making this comment once, but it applies to all with the same pattern.

Thalley replied:
Yes, the application can be both client and server. As per the API, setting `conn = NULL` will get the local server value, while `conn != NULL` will attempt to read the value from a remote server as the client. If this is unclear, we need to add more information in the documentation.

joerchan replied:
I think that is a bit unclear, my suggestion would be to have two functions.
Local values for the server should be `_get`/`_set`, and the clients `_read`/`_write` to show the GATT operation.
